File: baekjoon/1260.cpp
#include <iostream>
#include <queue>
#include <vector>
using namespace std;

void dfs(vector<vector<bool>> &v, vector<bool> &visit, int start){
		//불가능한 인덱스값(조건 만족하는 정점x) or 이미 방문한 노드이면
    if(start == -1 || visit[start]) return;

		//방문 처리 및 출력
    visit[start] = true;
    cout << start << " ";
    
    int s = v.size();
    int idx = -1; //다음에 방문할 인덱스값

    //방문할 점이 여러 개인 경우 정점 번호가 작은 것을 먼저 방문한다
    for(int i = 0;i < s;i++){
        if(v[start][i] && !visit[i]){
            idx = i;
            dfs(v, visit, idx);
        }
    }
}

void bfs(vector<vector<bool>> v, vector<bool> &visit, int start){
    queue<int> q; //다음에 방문할 노드 후보 집합
     q.push(-start); //큰 값부터 니오기 때문에 음수 활용

    while(!q.empty()){
        int s = v.size();
        int idx = -q.front(); q.pop();
        
        if(!visit[idx]){
            cout << idx << " ";
            visit[idx] = true;
        }
	    else continue;

		//간선이 연결되어 있고, 방문한 적 없는 노드라면 큐에 넣음
        for(int i = 0;i < s;i++)
            if(v[idx][i] && !visit[i]) q.push(-i);
    }
}

int main(){
    int n, m, start;
    int a, b; //입력받을 떄 사용할 임시 변수
    cin >> n >> m >> start;
    vector<vector<bool>> graph(n+1, vector<bool>(n+1, false)); //연결 여부만 저장한다
    vector<bool> visit(n+1, false); //방문했는가
  
    for(int i = 0;i < m;i++){
        cin >> a >> b;
        graph[a][b] = graph[b][a] = true;
    }

    dfs(graph, visit, start);
    cout << endl;
    fill(visit.begin(), visit.end(), false);
    bfs(graph, visit, start);
}

File: baekjoon/1260.java
package second;

import java.io.*;
import java.util.*;

public class BJ1260 {
	static StringBuffer sb = new StringBuffer();
	static int N, M, V;
	static boolean[] visited;
	static boolean[][] graph;
	
	static void dfs(int curNode) {
		if(visited[curNode]) return;
		
		visited[curNode] = true;
		sb.append(curNode).append(" ");
		for (int i = 1; i <= N; i++) {
			if(graph[curNode][i]) 
				dfs(i);
		}
	}
	
	static void bfs(int root) {
		//왜 arraylist대신 linkedlist를 사용하는가?
		//https://devlog-wjdrbs96.tistory.com/246
		Queue<Integer> nextNode = new LinkedList<>();
		nextNode.add(root);
		
		while(!nextNode.isEmpty()) {
			int curNode = nextNode.poll();
			if(visited[curNode]) continue;
			
			visited[curNode] = true;
			sb.append(curNode).append(" ");
			for (int i = 1; i <= N; i++) {
				if(graph[curNode][i]) 
					nextNode.add(i);
			}
		}
	}
	
	public static void main(String[] args) throws IOException {
		BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
		StringTokenizer st = new StringTokenizer(br.readLine());
		N = Integer.parseInt(st.nextToken());
		M = Integer.parseInt(st.nextToken());
		V = Integer.parseInt(st.nextToken());
		
		//초기화 작업
		visited = new boolean[N+1];
		graph = new boolean[N+1][N+1];
		
		//간선 정보 저장
		for (int i = 0; i < M; i++) {
			String edgeStr = br.readLine();
			int node1 = Integer.parseInt(edgeStr.split(" ")[0]);
			int node2 = Integer.parseInt(edgeStr.split(" ")[1]);
			graph[node1][node2] = true;		
			graph[node2][node1] = true;		
		}
		
		//탐색 시작
		dfs(V);
		Arrays.fill(visited, false);
		sb.append("\n");
		bfs(V);
		System.out.println(sb);
	}
}
